server/resp/src/main/java/org/infinispan/server/resp/commands/hash/HRANDFIELD.java

```diff
@@ -0,0 +1,98 @@
+package org.infinispan.server.resp.commands.hash;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+
+import org.infinispan.multimap.impl.EmbeddedMultimapPairCache;
+import org.infinispan.server.resp.ByteBufferUtils;
+import org.infinispan.server.resp.Consumers;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespErrorUtil;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.Util;
+import org.infinispan.server.resp.commands.ArgumentUtils;
+import org.infinispan.server.resp.commands.Resp3Command;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * `<code>HRANDFIELD key [count [WITHVALUES]]</code>` command.
+ * <p>
+ * Returns <code>count</code> keys from the hash stored at <code>key</code>. Negative <code>count</code> can return
+ * duplicated keys.
+ *
+ * @since 15.0
+ * @see <a href="https://redis.io/commands/hrandfield/">Redis documentation</a>
+ * @author José Bolina
+ */
+public class HRANDFIELD extends RespCommand implements Resp3Command {
+
+   private static final byte[] WITH_VALUES = "WITHVALUES".getBytes(StandardCharsets.US_ASCII);
+
+   public HRANDFIELD() {
+      super(-2, 1, 1, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      byte[] key = arguments.get(0);
+      int count = 1;
+      if (arguments.size() > 1) {
+         count = ArgumentUtils.toInt(arguments.get(1));
+      }
+
+      if (count == 0) {
+         ByteBufferUtils.bytesToResult(Collections.emptyList(), handler.allocator());
+         return handler.myStage();
+      }
+
+      boolean withValues = false;
+      if (arguments.size() > 2) {
+         // Syntax error, only a single option acceptable.
+         if (!Util.isAsciiBytesEquals(WITH_VALUES, arguments.get(2))) {
+            RespErrorUtil.syntaxError(handler.allocator());
+            return handler.myStage();
+         }
+         withValues = true;
+      }
+
+      EmbeddedMultimapPairCache<byte[], byte[], byte[]> multimap = handler.getHashMapMultimap();
+      int finalCount = count;
+      boolean finalWithValues = withValues;
+      CompletionStage<Collection<byte[]>> cs = multimap.subSelect(key, Math.abs(count))
```

Comment:
One thing I worry is that repeated invocations of a count < map size will always return the same subset of keys just in random orders. I don't think that is quite correct though.

Reply:
Updated to wrap the entry Set in a List and use the Collections.shuffle.